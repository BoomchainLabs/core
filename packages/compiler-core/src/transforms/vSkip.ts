import {
  ElementTypes,
  type IfBranchNode,
  NodeTypes,
  type SimpleExpressionNode,
  type SlotsExpression,
  type TemplateChildNode,
  createConditionalExpression,
  createSimpleExpression,
} from '../ast'
import {
  type NodeTransform,
  createStructuralDirectiveTransform,
} from '../transform'
import {
  ErrorCodes,
  createCompilerError,
  findProp,
  isSlotOutlet,
  isTemplateNode,
  processExpression,
} from '@vue/compiler-core'
import { createCodegenNodeForBranch } from './vIf'
import { validateBrowserExpression } from '../validateExpression'

export const transformSkip: NodeTransform = createStructuralDirectiveTransform(
  'skip',
  (node, dir, context) => {
    const loc = dir.exp ? dir.exp.loc : node.loc
    if (isTemplateNode(node) || isSlotOutlet(node)) {
      context.onError(createCompilerError(ErrorCodes.X_V_SKIP_ON_TEMPLATE, loc))
      return
    }

    if (!dir.exp || !(dir.exp as SimpleExpressionNode).content.trim()) {
      context.onError(
        createCompilerError(ErrorCodes.X_V_SKIP_NO_EXPRESSION, loc),
      )
      dir.exp = createSimpleExpression(`true`, false, loc)
    }

    if (!__BROWSER__ && context.prefixIdentifiers && dir.exp) {
      dir.exp = processExpression(dir.exp as SimpleExpressionNode, context)
    }

    if (__DEV__ && __BROWSER__ && dir.exp) {
      validateBrowserExpression(dir.exp as SimpleExpressionNode, context)
    }

    return () => {
      let children: TemplateChildNode[] = []
      // for components, extract default slot without props
      // if not found, throw an error
      if (node.tagType === ElementTypes.COMPONENT) {
        const codegenNode = node.codegenNode!
        if (codegenNode.type === NodeTypes.VNODE_CALL) {
          const genChildren = codegenNode.children! as SlotsExpression
          if (genChildren.type === NodeTypes.JS_OBJECT_EXPRESSION) {
            const prop = genChildren.properties.find(
              p =>
                p.type === NodeTypes.JS_PROPERTY &&
                p.key.type === NodeTypes.SIMPLE_EXPRESSION &&
                p.key.content === 'default' &&
                p.value.params === undefined,
            )
            if (prop) {
              children = prop.value.returns as TemplateChildNode[]
            } else {
              context.onError(
                createCompilerError(ErrorCodes.X_V_SKIP_UNEXPECTED_SLOT, loc),
              )
            }
          }
        }
      }
      // for plain elements, take all children
      else {
        children = node.children
      }
      const consequent: IfBranchNode = {
        type: NodeTypes.IF_BRANCH,
        loc: node.loc,
        condition: undefined,
        children,
        userKey: findProp(node, `key`),
      }

      const alternate: IfBranchNode = {
        type: NodeTypes.IF_BRANCH,
        loc: node.loc,
        condition: undefined,
        children: [node],
        userKey: findProp(node, `key`),
      }

      node.codegenNode = createConditionalExpression(
        dir.exp!,
        createCodegenNodeForBranch(consequent, 0, context),
        createCodegenNodeForBranch(alternate, 1, context),
      )
    }
  },
)
