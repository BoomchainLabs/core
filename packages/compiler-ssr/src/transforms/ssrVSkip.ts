import {
  ElementTypes,
  type NodeTransform,
  NodeTypes,
  type SkipNode,
  createCallExpression,
  createIfStatement,
  createStructuralDirectiveTransform,
  processSkip,
} from '@vue/compiler-core'
import { processIfBranch } from './ssrVIf'
import type { SSRTransformContext } from '../ssrCodegenTransform'
import { SSR_RENDER_SKIP_COMPONENT } from '../runtimeHelpers'

export const ssrTransformSkip: NodeTransform =
  createStructuralDirectiveTransform('skip', (node, dir, context) => {
    processSkip(node, dir, context)
    return () => {
      if (node.tagType === ElementTypes.COMPONENT && node.ssrCodegenNode) {
        const { arguments: args, loc } = node.ssrCodegenNode
        node.ssrCodegenNode = createCallExpression(
          context.helper(SSR_RENDER_SKIP_COMPONENT),
          [`_push`, dir.exp!, ...args],
          loc,
        )
      }
    }
  })

export function ssrProcessSkip(
  node: SkipNode,
  context: SSRTransformContext,
): void {
  const { consequent, alternate, test } = node

  // if consequent is an if branch, process it as well
  const consequentNode =
    consequent.type === NodeTypes.IF_BRANCH
      ? processIfBranch(consequent, context)
      : consequent

  const ifStatement = createIfStatement(
    test,
    consequentNode,
    processIfBranch(alternate, context),
  )
  context.pushStatement(ifStatement)
}
