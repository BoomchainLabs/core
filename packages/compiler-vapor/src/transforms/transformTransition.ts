import type { NodeTransform } from '@vue/compiler-vapor'
import { ElementTypes, NodeTypes } from '@vue/compiler-core'
import { isTransitionTag } from '../utils'
import { postTransformTransition } from '@vue/compiler-dom'

export const transformTransition: NodeTransform = (node, context) => {
  if (
    node.type === NodeTypes.ELEMENT &&
    node.tagType === ElementTypes.COMPONENT
  ) {
    if (isTransitionTag(node.tag)) {
      return postTransformTransition(node, context.options.onError)
    }
  }
}
