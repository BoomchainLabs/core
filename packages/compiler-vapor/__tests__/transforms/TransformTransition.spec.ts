import { makeCompile } from './_utils'
import {
  transformChildren,
  transformElement,
  transformText,
  transformVBind,
  transformVIf,
  transformVShow,
  transformVSlot,
} from '@vue/compiler-vapor'
import { transformTransition } from '../../src/transforms/transformTransition'
import { DOMErrorCodes } from '@vue/compiler-dom'

const compileWithElementTransform = makeCompile({
  nodeTransforms: [
    transformText,
    transformVIf,
    transformElement,
    transformVSlot,
    transformChildren,
    transformTransition,
  ],
  directiveTransforms: {
    bind: transformVBind,
    show: transformVShow,
  },
})

describe('compiler: transition', () => {
  test('basic', () => {
    const { code } = compileWithElementTransform(
      `<Transition appear><h1 v-show="show">foo</h1></Transition>`,
    )
    expect(code).toMatchSnapshot()
  })

  test('work with v-if', () => {
    const { code } = compileWithElementTransform(
      `<Transition><h1 v-if="show">foo</h1></Transition>`,
    )

    expect(code).toMatchSnapshot()
    // n2 should have a key
    expect(code).contains('n2.$key = 2')
  })

  test('work with dynamic keyed children', () => {
    const { code } = compileWithElementTransform(
      `<Transition>
        <h1 :key="key">foo</h1>
      </Transition>`,
    )

    expect(code).toMatchSnapshot()
    expect(code).contains('_createKeyedFragment(() => _ctx.key')
    // should preserve key
    expect(code).contains('n0.$key = _ctx.key')
  })

  test('warns if multiple children', () => {
    const onError = vi.fn()
    compileWithElementTransform(
      `<Transition>
        <h1>foo</h1>
        <h2>bar</h2>
      </Transition>`,
      {
        onError,
      },
    )
    expect(onError).toHaveBeenCalled()
    expect(onError.mock.calls).toMatchObject([
      [{ code: DOMErrorCodes.X_TRANSITION_INVALID_CHILDREN }],
    ])
  })

  test('inject persisted when child has v-show', () => {
    expect(
      compileWithElementTransform(`
        <Transition>
          <div v-show="ok" />
        </Transition>
    `).code,
    ).toMatchSnapshot()
  })

  // TODO more tests
})
