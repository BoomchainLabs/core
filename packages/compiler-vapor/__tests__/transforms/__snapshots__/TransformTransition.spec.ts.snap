// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transition > basic 1`] = `
"import { VaporTransition as _VaporTransition, applyVShow as _applyVShow, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, {
    appear: () => (""), 
    persisted: () => ("")
  }, {
    "default": () => {
      const n0 = t0()
      _applyVShow(n0, () => (_ctx.show))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > inject persisted when child has v-show 1`] = `
"import { VaporTransition as _VaporTransition, applyVShow as _applyVShow, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<div></div>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, { persisted: () => ("") }, {
    "default": () => {
      const n0 = t0()
      _applyVShow(n0, () => (_ctx.ok))
      return n0
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > work with dynamic keyed children 1`] = `
"import { VaporTransition as _VaporTransition, createKeyedFragment as _createKeyedFragment, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n1 = _createComponent(_VaporTransition, null, {
    "default": () => {
      return _createKeyedFragment(() => _ctx.key, () => {
        const n0 = t0()
        n0.$key = _ctx.key
        return n0
      })
    }
  }, true)
  return n1
}"
`;

exports[`compiler: transition > work with v-if 1`] = `
"import { VaporTransition as _VaporTransition, createIf as _createIf, createComponent as _createComponent, template as _template } from 'vue';
const t0 = _template("<h1>foo</h1>")

export function render(_ctx) {
  const n3 = _createComponent(_VaporTransition, null, {
    "default": () => {
      const n0 = _createIf(() => (_ctx.show), () => {
        const n2 = t0()
        n2.$key = 2
        return n2
      })
      return n0
    }
  }, true)
  return n3
}"
`;
